src/ui/customBreakpointsUI.ts

```diff
@@ -50,39 +74,86 @@ class BreakpointsDataProvider implements vscode.TreeDataProvider<Breakpoint> {
       }
 
       session.customRequest('enableCustomBreakpoints', {
-        ids: this.breakpoints.filter(b => b.enabled).map(b => b.id),
+        ids: this.breakpointCategories
+          .map(bC =>
+            bC.children.filter(b => b.checkboxState == vscode.TreeItemCheckboxState.Checked),
+          )
+          .flat(),
       });
     });
   }
 
-  getTreeItem(item: Breakpoint): vscode.TreeItem {
-    return item.treeItem;
+  getTreeItem(item: Breakpoint | BreakpointCategory): vscode.TreeItem {
+    return item;
   }
 
-  async getChildren(item?: Breakpoint): Promise<Breakpoint[]> {
-    if (!item) return this.breakpoints.filter(b => b.enabled).sort(Breakpoint.compare);
+  getChildren(
+    item?: Breakpoint | BreakpointCategory,
+  ): vscode.ProviderResult<(BreakpointCategory | Breakpoint)[]> {
+    if (!item) return this.breakpointCategories.sort(BreakpointCategory.compare);
+    if (item instanceof BreakpointCategory) {
+      return item.children.sort(Breakpoint.compare);
+    }
     return [];
   }
 
-  async getParent(): Promise<Breakpoint | undefined> {
+  getParent(item: Breakpoint | BreakpointCategory): vscode.ProviderResult<BreakpointCategory> {
+    if (item instanceof Breakpoint) {
+      return this.breakpointCategories.find(b => b.group === item.label);
+    }
     return undefined;
   }
 
   addBreakpoints(breakpoints: Breakpoint[]) {
-    for (const breakpoint of breakpoints) breakpoint.enabled = true;
+    if (breakpoints.length == 0) return;
+
+    for (const breakpoint of breakpoints)
+      breakpoint.checkboxState = vscode.TreeItemCheckboxState.Checked;
     const ids = breakpoints.map(b => b.id);
     for (const session of this._debugSessionTracker.getConcreteSessions())
       session.customRequest('enableCustomBreakpoints', { ids });
+
+    this.fixCheckState();
     this._onDidChangeTreeData.fire(undefined);
   }
 
-  removeBreakpoints(breakpointIds: CustomBreakpointId[]) {
-    const ids = new Set(breakpointIds);
-    for (const breakpoint of this.breakpoints) {
-      if (ids.has(breakpoint.id)) breakpoint.enabled = false;
+  removeBreakpoints(breakpoints: Breakpoint[]) {
+    if (breakpoints.length == 0) return;
+
+    for (const bC of this.breakpointCategories) {
+      if (bC.children.some(b => breakpoints.includes(b))) {
+        bC.checkboxState = vscode.TreeItemCheckboxState.Unchecked;
+      }
     }
+
+    // disable every breakpoint in breakpoints
+    for (const breakpoint of breakpoints)
+      breakpoint.checkboxState = vscode.TreeItemCheckboxState.Unchecked;
+
     for (const session of this._debugSessionTracker.getConcreteSessions())
-      session.customRequest('disableCustomBreakpoints', { ids: breakpointIds });
+      session.customRequest('disableCustomBreakpoints', { ids: breakpoints.map(e => e.id) });
+
+    this.fixCheckState();
```

Comment:
I don't think you need to do this here, you already handle unchecking parents on L123

Reply:
True. I put it there just in case